src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -0,0 +1,607 @@
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.h"
+#include "seastar/core/thread.hh"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore::omap_manager {
+std::ostream &operator<<(std::ostream &out, const omap_inner_key_t &rhs)
+{
+  return out << "omap_inner_key (" << rhs.key_off<< " - " << rhs.key_len
+             << " - " << rhs.laddr << ")";
+}
+
+std::ostream &operator<<(std::ostream &out, const omap_leaf_key_t &rhs)
+{
+  return out << "omap_leaf_key_t (" << rhs.key_off<< " - " << rhs.key_len
+             << " "<< rhs.val_off<<" - " << rhs.val_len << ")";
+}
+
+std::ostream &OMapInnerNode::print_detail_l(std::ostream &out) const
+{
+  return out << ", size=" << get_size()
+	     << ", depth=" << get_meta().depth;
+}
+
+OMapInnerNode::get_value_ret
+OMapInnerNode::get_value(omap_context_t oc, const std::string &key)
+{
+  logger().debug("{}: {} key = {}", "OMapInnerNode", __func__, key);
+  auto child_pt = get_containing_child(key);
+  auto laddr = child_pt->get_node_key().laddr;
+  return omap_load_extent(oc, this, laddr, get_meta().depth - 1).safe_then(
+    [oc, &key] (auto extent) {
+    return extent->get_value(oc, key);
+  });
+}
+
+OMapInnerNode::insert_ret
+OMapInnerNode::insert(omap_context_t oc, const std::string &key, const std::string &value)
+{
+  logger().debug("{}: {}  {}->{}", "OMapInnerNode",  __func__, key, value);
+  auto child_pt = get_containing_child(key);
+  assert(child_pt != iter_end());
+  auto laddr = child_pt->get_node_key().laddr;
+  return omap_load_extent(oc, this, laddr, get_meta().depth - 1).safe_then(
+    [this, oc, child_pt, &key, &value] (auto extent) {
+    auto size = extent->get_node_meta().depth > 1? key.size() + 1:
+                                           key.size() + value.size() + 2;
+    return extent->extent_is_overflow(size) ?
+      split_entry(oc, key, child_pt, extent) :
+      insert_ertr::make_ready_future<OMapNodeRef>(std::move(extent));
+  }).safe_then([oc, &key, &value](OMapNodeRef extent) {
+    return extent->insert(oc, key, value);
+  });
+}
+
+OMapInnerNode::rm_key_ret
+OMapInnerNode::rm_key(omap_context_t oc, const std::string &key)
+{
+  logger().debug("{}: {}","OMapInnerNode",  __func__);
+  auto child_pt = get_containing_child(key);
+  auto laddr = child_pt->get_node_key().laddr;
+  return omap_load_extent(oc, this, laddr, get_meta().depth - 1).safe_then(
+    [this, oc, &key, child_pt] (auto extent) {
+    if (extent->extent_under_median() && get_node_size() > 1) {
+      return merge_entry(oc, key, child_pt, extent);
+    } else {
+      return merge_entry_ertr::make_ready_future<OMapNodeRef>(std::move(extent));
+    }
+  }).safe_then([oc, &key](OMapNodeRef extent) mutable {
+    return extent->rm_key(oc, key);
+  });
+
+
+}
+
+OMapInnerNode::list_keys_ret
+OMapInnerNode::list_keys(omap_context_t oc, std::string start, size_t max_result_size)
+{
```

Comment:
This method and every other read in this file needs to hold a reference to *this until the return future resolves.  See 513ec5333b66fd1b5ded2ed9565371d4448cfc56 for an example.  If you use a normal transaction, you won't notice a problem because the transaction maintains a reference.  However, a weak_transaction will not and the caller will release the reference to *this before most of the code here actually completes.

Reply:
@athanatos I don't really understand weak_transaction here. do you mean I need capture "this" in every safe_then until finally return?

Reply:
The refcount on *this must be kept above zero until the end.  See the commit I referenced above, adding a safe_then at the holding a Ref by value should suffice.